### PR TITLE
swarm - feat: implement factory pattern with base and agent factories;

### PR DESCRIPTION
**Pull request by MichaelDecent**

### CODE REVIEW

**Comment by MichaelDecent** [comment]
@cobycloud i will appreciate your feedback.

**Review by cobycloud — CHANGES_REQUESTED** [review]
this sets multiple TypeVar as T

File: pkgs/swarmauri/tests/unit/factories/AgentFactory_unit_test.py
Comment (cobycloud) [inline pkgs/swarmauri/tests/unit/factories/AgentFactory_unit_test.py:1]: looks good

File: pkgs/swarmauri/tests/unit/factories/Factory_unit_test.py
Comment (cobycloud) [inline pkgs/swarmauri/tests/unit/factories/Factory_unit_test.py:1]: looks good

File: pkgs/swarmauri/swarmauri/factories/concrete/Factory.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    Non-recursive factory extending FactoryBase.
+    """
+    type: Literal['Factory'] = 'Factory'
+    _resource_registry: Dict[str, Dict[str, Type[T]]] = {}
+
+    def register(self, resource: str, type: str, resource_class: Type[T]) -> None:
+        """
+        Register a resource class under a specific resource and type.
+        """
+        if resource not in self._resource_registry:
```
Comment (cobycloud) [inline pkgs/swarmauri/swarmauri/factories/concrete/Factory.py]: this should auto-register all subclasses for all concrete objects in the swarmauri package.

we can add all the arrays from all of the concrete/__init__.py files to the registry.

**Review by cobycloud — CHANGES_REQUESTED** [review]

**Comment by MichaelDecent** [comment]
Noted!

File: pkgs/swarmauri/swarmauri/factories/concrete/Factory.py (hunk 1 above)
Comment (MichaelDecent) [reply pkgs/swarmauri/swarmauri/factories/concrete/Factory.py]: That way users don't need to register a subclass manually, but only create the specific object type.

**Comment by MichaelDecent** [comment]
Can we make class-specific factory have its related list of sub-classes already registered, that way you are only expected to create the an object instance from the list. What do you think ? @cobycloud

**Comment by cobycloud** [comment]
> Can we make class-specific factory have its related list of sub-classes already registered, that way you are only expected to create the an object instance from the list. What do you think ? @cobycloud

There should be a registration method so that anyone can bring their own class and register it. At the same time, we should auto-register any subclasses in the swarmauri package to the factory. We can do this by leveraging the `__all__ ` found in each of the `swarmauri/*/concrete/__init__.py` files

File: pkgs/swarmauri/swarmauri/factories/concrete/Factory.py (hunk 1 above)
Comment (cobycloud) [reply pkgs/swarmauri/swarmauri/factories/concrete/Factory.py]:
```python
    def create(self, resource: str, type: str, *args: Any, **kwargs: Any) -> T:
        """
        Create an instance of the class associated with the given resource and type.
        """
        if resource not in self._resource_registry:
            raise ValueError(f"Resource '{resource}' is not registered.")

        if type not in self._resource_registry[resource]:
            raise ValueError(
                f"Type '{type}' is not registered under resource '{resource}'."
            )

        cls = self._resource_registry[resource][type]
        return cls(*args, **kwargs)
```

Based on this, you are requiring some registration of a resource to the registry before the class would be able to create it for the user.

**Comment by cobycloud** [comment]
in the `swarmauri/*/concret/__init__.py` files, we can find logic as such:

```python
# File: Swarmauri/distances/concrete/__init__.py

import importlib

# Define a lazy loader function with a warning message if the module is not found
def _lazy_import(module_name, module_description=None):
    try:
        return importlib.import_module(module_name)
    except ImportError:
        # If module is not available, print a warning message
        print(f"Warning: The module '{module_description or module_name}' is not available. "
              f"Please install the necessary dependencies to enable this functionality.")
        return None

# List of distance names (file names without the ".py" extension)
distance_files = [
    "CanberraDistance",
    "ChebyshevDistance",
    "ChiSquaredDistance",
    "CosineDistance",
    "EuclideanDistance",
    "HaversineDistance",
    "JaccardIndexDistance",
    "LevenshteinDistance",
    "ManhattanDistance",
    "MinkowskiDistance",
    "SorensenDiceDistance",
    "SquaredEuclideanDistance",
]

# Lazy loading of distance modules, storing them in variables
for distance in distance_files:
    globals()[distance] = _lazy_import(f"swarmauri.distances.concrete.{distance}", distance)

# Adding the lazy-loaded distance modules to __all__
__all__ = distance_files

```

We can leverage `__all__` to register types of a resource kind to the Factory.

**Comment by MichaelDecent** [comment]
> in the `swarmauri/*/concret/__init__.py` files, we can find logic as such:
> 
> ```python
> # File: Swarmauri/distances/concrete/__init__.py
> 
> import importlib
> 
> # Define a lazy loader function with a warning message if the module is not found
> def _lazy_import(module_name, module_description=None):
>     try:
>         return importlib.import_module(module_name)
>     except ImportError:
>         # If module is not available, print a warning message
>         print(f"Warning: The module '{module_description or module_name}' is not available. "
>               f"Please install the necessary dependencies to enable this functionality.")
>         return None
> 
> # List of distance names (file names without the ".py" extension)
> distance_files = [
>     "CanberraDistance",
>     "ChebyshevDistance",
>     "ChiSquaredDistance",
>     "CosineDistance",
>     "EuclideanDistance",
>     "HaversineDistance",
>     "JaccardIndexDistance",
>     "LevenshteinDistance",
>     "ManhattanDistance",
>     "MinkowskiDistance",
>     "SorensenDiceDistance",
>     "SquaredEuclideanDistance",
> ]
> 
> # Lazy loading of distance modules, storing them in variables
> for distance in distance_files:
>     globals()[distance] = _lazy_import(f"swarmauri.distances.concrete.{distance}", distance)
> 
> # Adding the lazy-loaded distance modules to __all__
> __all__ = distance_files
> ```
> 
> We can leverage `__all__` to register types of a resource kind to the Factory.

Well noted!

**Comment by MichaelDecent** [comment]
[Issue 894](https://github.com/swarmauri/swarmauri-sdk/issues/894)

File: pkgs/swarmauri/swarmauri/factories/concrete/__init__.py
Comment (cobycloud) [inline pkgs/swarmauri/swarmauri/factories/concrete/__init__.py:1]: update this file, follow the examples set by the other `concrete/__init__.py` files

Comment (MichaelDecent) [reply pkgs/swarmauri/swarmauri/factories/concrete/__init__.py:1]: well noted!

**Review by cobycloud — APPROVED** [review]

**Pull request merged**